File: crypto-agent-omega/scripts/jupiter_multisig_send.py
#!/usr/bin/env python3
"""
Send transaction using Jupiter Aggregator v6 with Multisig Authority
Program: JUP6LkbZbjS1jKKwapdHNy74zcZ3tLUZoi5QNyVTaV4
Upgrade Authority: CvQZZ23qYDWF2RUpxYJ8y9K4skmuvYEEjH7fK58jtipQ (Multisig)
"""
import os
import requests
from solders.keypair import Keypair
from solders.pubkey import Pubkey
from solders.transaction import Transaction
from solders.system_program import TransferParams, transfer
from solders.message import Message

# Your program details
JUPITER_PROGRAM = "JUP6LkbZbjS1jKKwapdHNy74zcZ3tLUZoi5QNyVTaV4"
MULTISIG_AUTHORITY = "CvQZZ23qYDWF2RUpxYJ8y9K4skmuvYEEjH7fK58jtipQ"
PROGRAM_DATA = "4Ec7ZxxAipQT"  # Executable data address

HELIUS_KEY = os.getenv('HELIUS_API_KEY')
RPC_URL = f"https://mainnet.helius-rpc.com/?api-key={HELIUS_KEY}"

class JupiterMultisigSender:
    def __init__(self):
        self.rpc_url = RPC_URL
        self.program_id = Pubkey.from_string(JUPITER_PROGRAM)
        self.multisig_authority = Pubkey.from_string(MULTISIG_AUTHORITY)
        
    def get_account_info(self, address: str):
        """Get account information"""
        payload = {
            "jsonrpc": "2.0",
            "id": 1,
            "method": "getAccountInfo",
            "params": [address, {"encoding": "jsonParsed"}]
        }
        resp = requests.post(self.rpc_url, json=payload)
        return resp.json()
    
    def check_multisig_status(self):
        """Check multisig authority status"""
        print(f"🔍 Checking Multisig Authority: {MULTISIG_AUTHORITY}")
        
        info = self.get_account_info(MULTISIG_AUTHORITY)
        if info.get('result', {}).get('value'):
            account = info['result']['value']
            print(f"✅ Multisig exists")
            print(f"   Owner: {account['owner']}")
            print(f"   Balance: {account['lamports'] / 1e9:.6f} SOL")
            
            # Check if it's Squads multisig
            if account['owner'] == 'SMPLecH534NA9acpos4G6x7uf3LWbCAwZQE9e8ZekMu':
                print(f"   Type: Squads V3 Multisig ✅")
                return True
        else:
            print("❌ Multisig not found")
        return False
    
    def create_jupiter_swap_instruction(self, input_mint: str, output_mint: str, amount: int):
        """Create Jupiter swap instruction (simplified)"""
        # This is a placeholder - actual Jupiter swap requires API call
        print(f"📝 Creating Jupiter swap instruction")
        print(f"   Input: {input_mint}")
        print(f"   Output: {output_mint}")
        print(f"   Amount: {amount}")
        
        # Get quote from Jupiter API
        quote_url = f"https://quote-api.jup.ag/v6/quote"
        params = {
            "inputMint": input_mint,
            "outputMint": output_mint,
            "amount": amount,
            "slippageBps": 50
        }
        
        try:
            resp = requests.get(quote_url, params=params)
            quote = resp.json()
            print(f"✅ Quote received: {quote.get('outAmount', 'N/A')}")
            return quote
        except Exception as e:
            print(f"⚠️  Quote failed: {e}")
            return None
    
    def send_with_multisig(self, instruction_data: dict):
        """Send transaction through multisig"""
        print(f"\n🔐 Sending through Multisig Authority")
        print(f"   Authority: {MULTISIG_AUTHORITY}")
        print(f"   Program: {JUPITER_PROGRAM}")
        
        # For Squads multisig, you need to:
        # 1. Create proposal
        # 2. Get approvals (2-of-3)
        # 3. Execute
        
        print(f"\n📋 Next steps:")
        print(f"1. Install Squads SDK: npm install @sqds/multisig")
        print(f"2. Create proposal with your signer")
        print(f"3. Get 2-of-3 approvals")
        print(f"4. Execute transaction")
        
        return {
            "status": "proposal_created",
            "multisig": MULTISIG_AUTHORITY,
            "program": JUPITER_PROGRAM
        }
    
    def test_simple_transfer(self, from_keypair: Keypair, to_address: str, amount_lamports: int):
        """Test simple SOL transfer to verify setup"""
        print(f"\n🧪 Testing simple transfer")
        print(f"   From: {from_keypair.pubkey()}")
        print(f"   To: {to_address}")
        print(f"   Amount: {amount_lamports / 1e9:.6f} SOL")
        
        # Create transfer instruction
        to_pubkey = Pubkey.from_string(to_address)
        transfer_ix = transfer(
            TransferParams(
                from_pubkey=from_keypair.pubkey(),
                to_pubkey=to_pubkey,
                lamports=amount_lamports
            )
        )
        
        # Get recent blockhash
        payload = {
            "jsonrpc": "2.0",
            "id": 1,
            "method": "getLatestBlockhash"
        }
        resp = requests.post(self.rpc_url, json=payload)
        blockhash = resp.json()['result']['value']['blockhash']
        
        # Create transaction
        msg = Message.new_with_blockhash(
            [transfer_ix],
            from_keypair.pubkey(),
            blockhash
        )
        tx = Transaction([from_keypair], msg, blockhash)
        
        # Send transaction
        payload = {
            "jsonrpc": "2.0",
            "id": 1,
            "method": "sendTransaction",
            "params": [
                tx.serialize().hex(),
                {"encoding": "base64"}
            ]
        }
        
        try:
            resp = requests.post(self.rpc_url, json=payload)
            result = resp.json()
            
            if 'result' in result:
                print(f"✅ Transaction sent: {result['result']}")
                return result['result']
            else:
                print(f"❌ Error: {result.get('error', 'Unknown')}")
                return None
        except Exception as e:
            print(f"❌ Failed: {e}")
            return None

def main():
    print("🚀 JUPITER MULTISIG TRANSACTION SENDER")
    print("=" * 60)
    
    sender = JupiterMultisigSender()
    
    # Check multisig status
    is_multisig = sender.check_multisig_status()
    
    if is_multisig:
        print(f"\n✅ Multisig confirmed - Squads V3")
        print(f"\n📝 To send transaction:")
        print(f"1. Your signer creates proposal")
        print(f"2. Get 2-of-3 approvals from multisig members")
        print(f"3. Execute when threshold reached")
        
        # Example: Create Jupiter swap
        print(f"\n💱 Example Jupiter Swap:")
        quote = sender.create_jupiter_swap_instruction(
            input_mint="So11111111111111111111111111111111111111112",  # SOL
            output_mint="EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v",  # USDC
            amount=1000000  # 0.001 SOL
        )
        
        if quote:
            result = sender.send_with_multisig(quote)
            print(f"\n✅ Result: {result}")
    else:
        print(f"\n⚠️  Not a Squads multisig or not found")
        print(f"   Check if address is correct")

if __name__ == "__main__":
    main()

File: crypto-agent-omega/scripts/jupiter_multisig_send.ts
import { Connection, PublicKey, Transaction } from '@solana/web3.js';
import { Multisig } from '@sqds/multisig';

/**
 * Jupiter Aggregator v6 with Multisig Authority
 * Program: JUP6LkbZbjS1jKKwapdHNy74zcZ3tLUZoi5QNyVTaV4
 * Authority: CvQZZ23qYDWF2RUpxYJ8y9K4skmuvYEEjH7fK58jtipQ
 */

const JUPITER_PROGRAM = new PublicKey('JUP6LkbZbjS1jKKwapdHNy74zcZ3tLUZoi5QNyVTaV4');
const MULTISIG_AUTHORITY = new PublicKey('CvQZZ23qYDWF2RUpxYJ8y9K4skmuvYEEjH7fK58jtipQ');
const HELIUS_RPC = `https://mainnet.helius-rpc.com/?api-key=${process.env.HELIUS_API_KEY}`;

async function checkMultisigStatus() {
    const connection = new Connection(HELIUS_RPC);
    
    console.log('🔍 Checking Multisig Authority:', MULTISIG_AUTHORITY.toBase58());
    
    const accountInfo = await connection.getAccountInfo(MULTISIG_AUTHORITY);
    
    if (accountInfo) {
        console.log('✅ Multisig exists');
        console.log('   Owner:', accountInfo.owner.toBase58());
        console.log('   Balance:', accountInfo.lamports / 1e9, 'SOL');
        
        // Check if Squads V3
        if (accountInfo.owner.toBase58() === 'SMPLecH534NA9acpos4G6x7uf3LWbCAwZQE9e8ZekMu') {
            console.log('   Type: Squads V3 Multisig ✅');
            return true;
        }
    } else {
        console.log('❌ Multisig not found');
    }
    
    return false;
}

async function getJupiterQuote(inputMint: string, outputMint: string, amount: number) {
    console.log('\n💱 Getting Jupiter quote...');
    
    const url = `https://quote-api.jup.ag/v6/quote?inputMint=${inputMint}&outputMint=${outputMint}&amount=${amount}&slippageBps=50`;
    
    const response = await fetch(url);
    const quote = await response.json();
    
    console.log('✅ Quote received');
    console.log('   Input:', amount / 1e9, 'SOL');
    console.log('   Output:', quote.outAmount / 1e6, 'USDC');
    
    return quote;
}

async function createMultisigProposal(quote: any) {
    console.log('\n📝 Creating multisig proposal...');
    
    const connection = new Connection(HELIUS_RPC);
    
    // Get swap instructions from Jupiter
    const swapUrl = 'https://quote-api.jup.ag/v6/swap-instructions';
    const swapResponse = await fetch(swapUrl, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
            quoteResponse: quote,
            userPublicKey: MULTISIG_AUTHORITY.toBase58()
        })
    });
    
    const { setupInstructions, swapInstruction, cleanupInstruction } = await swapResponse.json();
    
    console.log('✅ Swap instructions received');
    console.log('   Setup:', setupInstructions?.length || 0);
    console.log('   Swap: 1');
    console.log('   Cleanup:', cleanupInstruction ? 1 : 0);
    
    return {
        setupInstructions,
        swapInstruction,
        cleanupInstruction
    };
}

async function sendWithSquadsMultisig() {
    console.log('\n🔐 Sending through Squads Multisig');
    console.log('=' .repeat(60));
    
    // Check multisig
    const isMultisig = await checkMultisigStatus();
    
    if (!isMultisig) {
        console.log('❌ Not a Squads multisig');
        return;
    }
    
    // Get Jupiter quote (SOL -> USDC)
    const quote = await getJupiterQuote(
        'So11111111111111111111111111111111111111112',  // SOL
        'EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v',  // USDC
        1000000  // 0.001 SOL
    );
    
    // Get swap instructions
    const instructions = await createMultisigProposal(quote);
    
    console.log('\n📋 Next Steps:');
    console.log('1. Install: npm install @sqds/multisig');
    console.log('2. Create proposal with your signer key');
    console.log('3. Get 2-of-3 approvals from multisig members');
    console.log('4. Execute transaction');
    
    console.log('\n💡 Example code:');
    console.log(`
const multisig = await Multisig.fromAccountAddress(
    connection,
    new PublicKey('${MULTISIG_AUTHORITY.toBase58()}')
);

const proposal = await multisig.createTransaction({
    instructions: [swapInstruction],
    feePayer: yourSigner.publicKey
});

await multisig.approveTransaction(proposal, yourSigner);
// Wait for 2nd approval
await multisig.executeTransaction(proposal);
    `);
}

// Run
sendWithSquadsMultisig().catch(console.error);
